compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,471 @@
+                        adt_def.all_fields().map(|field| field.ty(tcx, substs)).collect();
+                    if !tys.is_empty() {
+                        s.push('I');
+                        for ty in tys {
+                            s.push_str(&encode_ty(tcx, ty, dict, encode_c_types));
+                        }
+                        s.push('E');
+                    }
+                }
+                s = format!("u{}{}", s.len(), &s);
+                compress(dict, (ty, TyQ::None), &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::Foreign(def_id) => {
+            // <length><name>
+            let mut s = String::new();
+            let name = format!("{}", tcx.item_name(*def_id));
+            s.push_str(&format!("{}", name.len()));
+            s.push_str(&name);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Function types
+        ty::FnDef(def_id, ..) => {
+            // u2fnI<length><name>E as vendor extended type
+            let mut s = String::from("u2fnI");
+            s.push_str(&encode_ty_name(tcx, *def_id));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Closure(_, substs) | ty::Generator(_, substs, _) => {
+            // u7closureI<return-type><element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u7closureI");
+            let sig = substs.as_closure().sig();
+            // Push str without the initial 'F' but with the closing 'E'
+            s.push_str(&(encode_fnsig(tcx, &sig.skip_binder(), dict)[1..]));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Pointer types
+        ty::Ref(_, ty0, mutbl) => {
+            // [U3mut]u3refI<element-type>E as vendor extended type qualifier and type
+            let mut s = String::new();
+            s.push_str("u3refI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            if hir::Mutability::Mut == *mutbl {
+                s = format!("{}{}", "U3mut", &s);
+                compress(dict, (ty, TyQ::Mut), &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::RawPtr(tm) => {
+            // P[K]<element-type>
+            let mut s = String::new();
+            s.push_str(&encode_ty(tcx, tm.ty, dict, encode_c_types));
+            if hir::Mutability::Not == tm.mutbl {
+                s = format!("{}{}", "K", &s);
+                compress(dict, (tm.ty, TyQ::Const), &mut s);
+            };
+            s = format!("{}{}", "P", &s);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::FnPtr(sig) => {
+            // F<return-type><parameter-type1..parameter-typeN>E
```

Comment:
It seems `FnPtr` should be a pointer to a function, i.e., `<type> => P <type> => P <function-type> => P F <bare-function-type> E`

Reply:
Done.